packages/ckeditor5-core/src/plugincollection.js

```diff
@@ -163,35 +162,157 @@ export default class PluginCollection {
 	 * @returns {Promise.<module:core/plugin~LoadedPlugins>} A promise which gets resolved once all plugins are loaded
 	 * and available in the collection.
 	 */
-	init( plugins, removePlugins = [] ) {
+	init( plugins, removedPlugins = [] ) {
+		// Plugin initialization procedure consists of 2 main steps:
+		// 1) collecting all available plugin constructors,
+		// 2) verification whether all required plugins can be instantiated.
+
+		// In the first step, all plugin constructors, available in the provided `plugins` array and inside
+		// plugin's dependencies (from the `requires` array), are recursively collected and added to the existing
+		// `this._availablePlugins` map, but without any verification at the given moment. Performing the verification
+		// at this point (during the plugin constructor searching) would cause false errors to occur, that some plugin
+		// is missing but in fact it may be defined further in the array as the dependency of other plugin. After
+		// traversing the entire dependency tree, it will be checked if all required "top level" plugins are available.
+
+		// In the second step, the list of plugins that have not been explicitly removed is traversed to get all the
+		// plugin constructors to be instantiated in the correct order and to validate against some rules. Finally, if
+		// no plugin is missing and no other error has been found, they all will be instantiated.
 		const that = this;
 		const context = this._context;
-		const loading = new Set();
-		const loaded = [];
-		const loadedPluginsNames = new Set();
 
-		const pluginsMap = new Map();
+		findAvailablePluginConstructors( plugins );
 
-		for ( const pluginOrName of plugins ) {
-			if ( typeof pluginOrName == 'function' && pluginOrName.pluginName ) {
-				pluginsMap.set( pluginOrName.pluginName, pluginOrName );
-			}
+		validatePlugins( plugins );
+
+		const pluginsToLoad = plugins.filter( plugin => !isPluginRemoved( plugin, removedPlugins ) );
+
+		const pluginConstructors = [ ...getPluginConstructors( pluginsToLoad ) ];
+
+		const pluginInstances = loadPlugins( pluginConstructors );
+
+		return initPlugins( pluginInstances, 'init' )
+			.then( () => initPlugins( pluginInstances, 'afterInit' ) )
+			.then( () => pluginInstances );
+
+		function isPluginConstructor( plugin ) {
+			return typeof plugin === 'function';
+		}
+
+		function isContextPlugin( plugin ) {
+			return isPluginConstructor( plugin ) && plugin.isContextPlugin;
 		}
 
-		const pluginConstructors = mapToAvailableConstructors( plugins );
-		const removePluginConstructors = mapToAvailableConstructors( removePlugins );
-		const missingPlugins = getMissingPluginNames( plugins );
+		function isPluginRemoved( plugin, removedPlugins ) {
+			return removedPlugins.some( removedPlugin => {
+				return removedPlugin === plugin ||
+					removedPlugin === getPluginName( plugin ) ||
+					getPluginName( removedPlugin ) === plugin;
+			} );
+		}
```

Comment:
I'd refactor this function a bit:

```suggestion
		function isPluginRemoved( plugin, removedPlugins ) {
			return removedPlugins.some( removedPlugin => {
				if ( removedPlugin === plugin ) {
					return true;
				}
				
				if ( getPluginName( plugin ) === removedPlugin ) {
					return true;
				}

				if ( getPluginName( removedPlugin ) === plugin ) {
					return true;
				}
				
				return false;
			} );
		}
```

Reply:
Done.